### PR TITLE
L49: Objective-C Flow Control

### PR DESCRIPTION
**Pull request by muxi**
Discussion link: https://groups.google.com/forum/#!topic/grpc-io/xSOtIAKSESk

### CODE REVIEW

File: L49-objc-flow-control.md [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+
+@end
+```
+A user that enabled flow control is supposed to wait for issuance of
+`didWriteData` callback before making another `writeData` call. However, if
+`writeData` are called multiple times, e.g. due to a race to write messages,
+gRPC will still send all of the writes in order and issue `didWriteData` for
+each invoke of `writeData`. However in this case users assume their own
+responsibility of memory issue. When flow control is not enabled, the callback
+`didWriteData` is not issued to the user.
```
[the remark this answers is not in the document]
Comment (muxi) [reply L49-objc-flow-control.md]: Updated

File: L49-objc-flow-control.md [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+### Write flow control performance
+The write flow control implies only one message be buffered in gRPC Objective-C
+layer, and the `didWriteData` callback is issued to another thread. So a
+context switch happens for each write of message. On mobile platform this may
+not be a big issue, but in case some users do send a large amount of small
+messages and find the performance poorer than their expectation, the issue may
+be resolved in the following ways:
+* since gRPC does not put a hard restriction on how many messages, users may
+  self-restrict the number of pending writes without corresponding
+  `didWriteData` callback issued
```
[the remark this answers is not in the document]
Comment (muxi) [reply L49-objc-flow-control.md]: rephrased this part.

File: L49-objc-flow-control.md [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+* gRPC Objective-C layer may detect the size of message at `writeData` function
+  and immediately issue back a `didWriteData` callback if the message size is
+  small.
+
+### Write flow control API
+We may put a hard limit on the write flow control such that when more than one
```
[the remark this answers is not in the document]
Comment (muxi) [reply L49-objc-flow-control.md]: updated

File: L49-objc-flow-control.md [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+...
+
+@end
+```
+
+Both read and write flow control (described below) are enabled only if the boolean's value is YES.
+
+### Write flow control
+Write flow control intend to slow down user's message writes such that there's
+only one pending message in gRPC Objective-C layer. This is achieved by an
```
[the remark this answers is not in the document]
Comment (muxi) [reply L49-objc-flow-control.md]: updated

File: L49-objc-flow-control.md [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+/** Enable flow control feature */
+@property BOOL enableFlowControl;
+
+...
+
+@end
+```
+
+Both read and write flow control (described below) are enabled only if the boolean's value is YES.
```
[the remark this answers is not in the document]
Comment (muxi) [reply L49-objc-flow-control.md]: Good catch. I was not aware that there's a variable for the core too. Updated the doc.

[the remark this answers is not in the document]
Comment (srini100) [reply L49-objc-flow-control.md]: The formatting of GRPC\_EXPERIMENTAL\_DISABLE\_FLOW\_CONTROL is not looking right.

**Pull request merged**